### PR TITLE
Adds JavaScript guides

### PR DESCRIPTION
**Pull request by joaojusto**
Adds first version of JavaScript style guides

### CODE REVIEW

File: style/js/README.md [hunk 1]
```diff
@@ -1,3 +1,190 @@
...
+    //some work here
+  }
+
+  bar(foo, callback);
+```
+
+Conditional evaluation
+---
+
+* Always use `===`.
```
Comment (gabrielpoca) [inline style/js/README.md:122]: I think maybe we should have some reference to es6 in these guides.

File: style/js/README.md [hunk 2]
```diff
@@ -1,3 +1,190 @@
...
+  this.bar = bar;
+}
+
+Foo.prototype.getBar = function() {
+  return this.bar;
+};
+
+var foo = new Foo("bar");
+
+// or
```
Comment (gabrielpoca) [inline style/js/README.md]: The results of the two approaches are different.

File: style/js/README.md [hunk 3]
```diff
@@ -1,3 +1,190 @@
...
+* Prefer this form for multiline object/array initialization
+
+```javascript
+var map = {
+  when: 4,
+  ready: 9,
+  "you are": 15
+};
+
+var map = [
```
Comment (gabrielpoca) [inline style/js/README.md]: For a case like this one I don't think we should force this style.

Comment (joaojusto) [inline style/js/README.md]: What do you suggest then? The rule is not enforcing, its suggesting

File: style/js/README.md (hunk 2 above)
Comment (joaojusto) [inline style/js/README.md]: Should we break them in two cases then?

File: style/js/README.md (hunk 1 above)
Comment (joaojusto) [inline style/js/README.md:122]: What kind of references?

Comment (zamith) [inline style/js/README.md:122]: Are we using es6 on any project?

File: style/js/README.md [hunk 4]
```diff
@@ -1,3 +1,190 @@
+JavaScript Style Guides
+=====
+
+Guides for programing JavaScript in style
+
+General
+---
+* Use [JSHint](http://jshint.com/)
+* Always use 'use strict'
```
Comment (zamith) [inline style/js/README.md]: I think there some backticks missing here, around the use strict.

File: style/js/README.md (hunk 3 above)
Comment (zamith) [inline style/js/README.md]: For arrays I would say that it only applies if the resulting array declaration would be too long, more than 80 chars, for instance. At least that's how I usually do it, anyway.

File: style/js/README.md (hunk 2 above)
Comment (zamith) [inline style/js/README.md]: I'm not sure if this should even be in here. As we have discussed there are many ways of creating "classes" in javascript, and each has its merits. It's more of a case of choosing what's best for each scenario, than it is having a preferred way of doing it, I believe.

**Comment by zamith** [comment]
@joaojusto What about moving the examples to a separate file or files, and reference them from the main guide? It's what we are using for all the other guides, and makes it much easier to read, in my opinion.

File: style/js/README.md (hunk 1 above)
Comment (gabrielpoca) [inline style/js/README.md:122]: @zamith on one client project and on theshelf-angular. I believe @azevedo-252 and @ronaldofs agree with me, that we will try to move into it as soon as possible.

File: style/js/README.md (hunk 3 above)
Comment (gabrielpoca) [inline style/js/README.md]: @zamith :+1:  that's a good rule

Comment (naps62) [inline style/js/README.md]: not really. I often do this for short lines as well, when I want to add some visual structure. For instance, I just looked at my other monitor and noticed that I've written this:

``` javascript
var frontVertices = [
  0,     1 * rowHeight,
  width, 3 * rowHeight,
  width, 4 * rowHeight
];
```

In this example, each 2 values represents a point, and that visual organization helps a lot. And I do this a lot in other cases as well. I guess it depends a lot more on context rather than just the length of the line

PS: and please don't bother commenting on how `rowHeight` is duplicated, or each point should have their own object. That's beside the point, and is dependent on the API I'm using.

Comment (gabrielpoca) [inline style/js/README.md]: @naps62 yes, that's clearly a scenario for multiple lines. It's not the same as the example on the file.

Comment (zamith) [inline style/js/README.md]: @naps62 In that case I would probably wrap each point in an javascript object, so that I don't overload the array with the responsibility of knowing that structure. And only then would I break it into lines. Something like this:

``` javascript
var frontVertices = [
  { width: 0,     height: 1 * rowHeight },
  { width: width, height: 3 * rowHeight },
  { width: width, height: 4 * rowHeight}
];
```

But regardless of that, I don't feel strongly about any enforced style for this, even though I only break it when it doesn't fit on a line.

File: style/js/README.md (hunk 1 above)
Comment (zamith) [inline style/js/README.md:122]: If it is already being used, then it should be on the guides.

File: style/js/README.md (hunk 3 above)
Comment (joaojusto) [inline style/js/README.md]: @zamith  What you are saying might be a good solution if the API Miguel is using accepts it that way. Probably won't. 
Regarding the rule, when I wrote it I was thinking in suggesting a good way, not enforcing anything . |f its not clear maybe we can reword it

File: style/js/README.md (hunk 4 above)
Comment (joaojusto) [inline style/js/README.md]: What kind of backtricks?

**Comment by joaojusto** [comment]
@zamith I know the other guides have the samples in a separate file but I did them in line on purpose and the feedback from the people I talk was good. But let them say something and then if we all prefer the other way I'll change.

File: style/js/README.md (hunk 2 above)
Comment (joaojusto) [inline style/js/README.md]: In my opinion thats exactly why we should propose some of them. And notice that I said propose not enforce.

File: style/js/README.md (hunk 3 above)
Comment (naps62) [inline style/js/README.md]: @zamith I specifically added a note at the end of my previous comment about that, which you seem to have skipped:

> PS: and please don't bother commenting on how rowHeight is duplicated, or each point should have their own object. That's beside the point, and is dependent on the API I'm using.

Comment (zamith) [inline style/js/README.md]: I only saw the first part. Either way, this seems like a really really edge case. Is there any other example of something similar?

Comment (zamith) [inline style/js/README.md]: @joaojusto I understand that, and what I'm arguing is that the proposed way is not a good way. As a rule of thumb I'd prefer to have what I mentioned in my first comment.

I do agree that this should be a "Prefer" and not a "Use".

File: style/js/README.md (hunk 2 above)
Comment (zamith) [inline style/js/README.md]: What would be the benefit of proposing more than one? If we don't agree that one solution is the best, or at least the best for the vast majority of cases, then it has no business on the guides.

The guides are not a list of possible ways of doing something, but rather our preferred way of doing it.

File: style/js/README.md (hunk 3 above)
Comment (naps62) [inline style/js/README.md]: Now that I notice, we're discussing if we should prefer/suggest multiline arrays, but this section of the guides is not about that. It's about how to style a multiline array when we do use one:

> Prefer this form for multiline object/array initialization

So I guess we went completely offtopic here

File: style/js/README.md (hunk 2 above)
Comment (joaojusto) [inline style/js/README.md]: These two produce different results, so we need to clarify that and write them down. The guides in my opinion, aren't a set of rules that can't be broken, but more like a set of padrons that should be followed whenever possible. This two cases are a common padron and so to me it make senses to document them.

File: style/js/README.md (hunk 3 above)
Comment (joaojusto) [inline style/js/README.md]: Agree :+1:
BTW @zamith its already prefer.

File: style/js/README.md (hunk 2 above)
Comment (zamith) [inline style/js/README.md]: Since they produce different results, is one better than the other for different situations? If so, which?

Comment (ronaldofs) [inline style/js/README.md]: The problem here is that the 2nd version shouldn't even be used. Every time an instance of `Foo` is created, the methods are added to the instance. You only share methods thru `prototype`, so the 1st is correct.

Comment (zamith) [inline style/js/README.md]: If so, then let's say that. I'm good with it if we only propose one.

Comment (naps62) [inline style/js/README.md]: @ronaldofs that doesn't mean it shouldn't be used. It might have it's use cases, as long as you know what you're doing and the consequences. Of course, you probably do want to share the methods most of the times (i.e., use the first option), but saying we should never use the other is an exaggeration.

But getting back on topic:

> These two produce different results, so we need to clarify that and write them down

I don't agree with this at all. These guides are not a Javascript tutorial. They're not about explaining how different things work. If these two have different results, they should be used accordingly. Do we have any particular rules or restrictions about where to use either one? If so, **that** should be documented.

Right now, this has no explanation at all, it just says `Constructors` followed by an enumeration of both of them.
That clarifies nothing about their differences

File: style/js/README.md (hunk 3 above)
Comment (zamith) [inline style/js/README.md]: Right. Then I propose this to be accompanied with a "prefer single line arrays" rule.

Also, moving the examples to a separate file.

File: style/js/README.md (hunk 2 above)
Comment (gabrielpoca) [inline style/js/README.md]: @ronaldofs That's not true, I can think of a couple of situations where you need the second version. If you want to change the prototype without affecting other objects for instance. I think we can all agree that this doesn't belong here.

Comment (ronaldofs) [inline style/js/README.md]: @naps62 I believe this pattern isn't good and also I couldn't come up with a use case that couldn't be achieved with better approaches. So, my advice is not to use it. Ever. xD

Comment (ronaldofs) [inline style/js/README.md]: @gabrielpoca when do you need a class that duplicates instance methods and has the same constructor function? Is there an example? If you'd want to add behavior to specific instances, you could: use decorators, mixins or method override. But not in the constructor...

Comment (gabrielpoca) [inline style/js/README.md]: @ronaldofs there are multiple ways. But you may be dealing with external libraries. So I don't believe we have to be so strict as to say never.

Comment (zamith) [inline style/js/README.md]: If the only use case we can come up with not to use it, is "because 3rd party libs might not let us", they I think this is a case for a "Prefer", at the very least.

File: style/js/README.md (hunk 4 above)
Comment (zamith) [inline style/js/README.md]: `'use strict'`

**Comment by zamith** [comment]
@joaojusto I strongly disagree with this. The guides are meant to be read quickly, and if each point is well explained, understanding it should be pretty much straight forward. 

In situations where it might not be as easy to explain, then an example is in order. However, that should not be in the middle of other rules, since it damps the ease of reading.

File: style/js/README.md (hunk 3 above)
Comment (joaojusto) [inline style/js/README.md]: I agree with the rule, but I don't agree with moving the examples to a separate file.

**Comment by joaojusto** [comment]
@zamith I don't believe the example code in the middle of the rules will be causing any embarrassment to the reader, in fact I believe it's the other way around, all the JavaScript guides I've been reading have the examples inline and it was easier to me to understand the rule every single time. Also, I think, the amount of code is not enough to decrease the velocity that you can find a rule here.

Bottom line, I think we need input from more people in this matter.

File: style/js/README.md (hunk 2 above)
Comment (gabrielpoca) [inline style/js/README.md]: I didn't take the time to think about scenarios. But there are obviously other ways around every scenario I give. But ok I'll give you a real world situation where it's perfectly acceptable to use the second option: singleton. And yes, you can use the first option too, but the second one reads better, and has no disadvantage.

Comment (zamith) [inline style/js/README.md]: I have no particular opinion here, other than, if there is more than one accepted way of doing it, then it should be removed from the guides entirely, since as @naps62 said, this is not a tutorial.

So, if you and @ronaldofs can reach an agreement on this, I'm good with it as well.

**Comment by gabrielpoca** [comment]
@groupbuddies/developers bump

**Comment by zamith** [comment]
@joaojusto That is not the format of our guides though. Actually, not even the individual rules are phrased correctly, according to our guides. They should all use the words "Use", "Prefer", "Avoid" or "Don't".

**Comment by gabrielpoca** [comment]
@groupbuddies/developers bump

**Comment by joaojusto** [comment]
@zamith I can agree with you regarding the phrasing, but I already said that this format to me feels more readable than the previous. @gabrielpoca  @naps62 @azevedo-252 @ronaldofs what do you think? Should we move the samples to a separated file or should we leave them inline?

File: style/js/README.md (hunk 2 above)
Comment (ronaldofs) [inline style/js/README.md]: If the guides are more about pointing towards code style conventions then yes, examples should be removed.

**Comment by gabrielpoca** [comment]
I agree that it's not consistent with the rest of the guides, but maybe we need to change those.

I rather see the code inline, most times it's easier to understand looking at the examples. I can't remember another guide where the rules are in samples instead.

File: style/js/README.md [hunk 5]
```diff
@@ -1,3 +1,190 @@
...
+if (something) {
+  // do stuff
+  // even more stuff
+} else {
+  // some other stuff
+  // real stuff
+}
+```
+
+* On if statements, whenever it makes sense, you can omit the `{}`.
```
Comment (zamith) [inline style/js/README.md]: Why do we want this?

I can see some disadvantages:
1. We have more than one way of doing the same thing (if statements) and thus have to think which one to use. This is something where I believe no thinking should be involved.
2. If it grows to more than one line per branch, we need to change it, which is putting the "pain" forward to the next guy.
3. It fails jslint or jshint, one of those, I believe.

On the other hand, the only advantage I see is that we spare a few characters.

**Comment by zamith** [comment]
If you look through the examples (and disregarding the constructors part), you'll probably be able to see that out of all of them, only 3 are adding something to what the sentence is already saying. From these 3, I believe that in at least 2 the sentence can be reworded to make the examples redundant as well.

If you read through other of our guides, such as the Ruby ones, I think that if you are at all familiar with the language, you'll understand the guides without the extra noise of the samples, which would make them harder to read (if for nothing else, for having a lot more lines). I would argue that the samples there, aren't even needed at all, but they are there with all of the "rules" put together, to give the sense of how it all looks in a bigger example, and to disambiguate some possible situations.

**Comment by gabrielpoca** [comment]
I think the style guide isn't only for us. I guess that if this wasn't on github we would have a button to toggle the styles inline.

Here are two popular styleguides that I really like.

https://github.com/necolas/idiomatic-css
https://github.com/airbnb/javascript

**Comment by zamith** [comment]
Most of the examples on the airbnb styleguide are redundant and add nothing to the "rule", but it's their choice to keep it like that or change it.

I haven't had any problem with understanding any of the "rules" in our guides, and without the code examples they are much easier to read. If I ever do have any issue understand something, then my opinion might change, but I think all of the "rules" are pretty much straightforward and the code inline is just noise.

**Comment by gabrielpoca** [comment]
I think you argument is all about "I", and I don't believe that should be the focus since you probably aren't the most active vistor for this.

**Comment by zamith** [comment]
I am giving my opinion, I can only base it on my personal experience. Have you or anyone else had any issue understanding the other guides?

File: style/js/README.md (hunk 1 above)
Comment (gabrielpoca) [inline style/js/README.md:122]: My bad, in es6 you still need `===` even tough there is `Object.is`.

**Comment by gabrielpoca** [comment]
@subvisual/developers bump

**Comment by joaojusto** [comment]
@subvisual/developers Updated some rules where the discussion had ended.

**Comment by joaojusto** [comment]
@subvisual/developers bump

File: style/js/README.md [hunk 6]
```diff
@@ -1,3 +1,163 @@
...
+}
+
+var bar = function() {
+  return foo;
+}
+```
+
+Assignments and Declarations
+---
+* Initialize arrays with `[ ]`.
```
Comment (naps62) [inline style/js/README.md]: you ask to initialize with `[ ]`, and then in the example you use `[]`. Since this is a style guide, I'd say the spacing in there should be consistent, otherwise it will lead to confusion

File: style/js/README.md [hunk 7]
```diff
@@ -1,3 +1,163 @@
...
+
+```javascript
+var array = [];
+```
+
+* Initialize objects with `{ }`.
+
+```javascript
+var object = {};
```
Comment (naps62) [inline style/js/README.md:56]: same here

File: style/js/README.md [hunk 8]
```diff
@@ -1,3 +1,163 @@
...
+```
+
+* Leave one space after `{` and before `}` when initializing objects on the same line.
+
+```javascript
+var map = { ready: 9, when: 4, "you are": 15 };
+```
+
+* Prefer single line object/array initialization.
```
Comment (naps62) [inline style/js/README.md:65]: I don't think we actually follow this. Unless the object is really small, but that's more an edge case than the default rule I think

File: style/js/README.md (hunk 6 above)
Comment (joaojusto) [inline style/js/README.md]: My bad, this wasn't about the spacing but about using the `[]` instead of `new Array();`

File: style/js/README.md (hunk 7 above)
Comment (joaojusto) [inline style/js/README.md:56]: Same thing, I'll correct the spacing

File: style/js/README.md (hunk 8 above)
Comment (joaojusto) [inline style/js/README.md:65]: That's why it's a prefer. Obviously if the object it's big it should be broken into multiple lines.

File: style/js/README.md (hunk 6 above)
Comment (naps62) [inline style/js/README.md]: I know that, but it's better to avoid confusion

File: style/js/README.md (hunk 8 above)
Comment (naps62) [inline style/js/README.md:65]: then what's the point in asking to prefer single line initialization, when it actually should be "prefer single line as long as the object is small, otherwise prefer multiline" ?

Comment (joaojusto) [inline style/js/README.md:65]: The point is to avoid things like

``` javascript
var foo = {
  bar: true
};
```

when you could write like this:

``` javascript
var foo = { bar: true };
```

This way the rule is to use single line, and the exception is the multiline form. Don't you agree that there is a difference here?

Comment (naps62) [inline style/js/README.md:65]: of course there is a difference (although I disagree that inline should be preferred, and a quick look at the oncostyle code tells me I might not be the only one. But that's not the point here)

An exception is something exceptional (duh). Having large objects (i.e., multiline), it not something exceptional, we do it all the time. So this guideline will only actually apply to a small percentage of all objects

But again, I just found several objects not following this guide in our recent repos. So I have to ask if we actually want this here

Comment (joaojusto) [inline style/js/README.md:65]: You can ask the question the other way around, do we want the other form on our repos? Let the other developers pitch in too so we can understand what is really prefered by the majority instead of guessing around :walking:

Comment (gabrielpoca) [inline style/js/README.md:65]: I don't think the guides should be justified by the code that's already in the repos, but the code we will write in the future. I agree that small objects should be single-line, although it's hard to tell what's small.

Comment (naps62) [inline style/js/README.md:65]: I don't really see why we shouldn't look at how we write code in the present to understand how we want to write code in the future. If we find ourselves using a certain style, and having no problem with it, why change it? Just for the sake of changing?

Anyway, this is not really my place to discuss, and everyone else seems to agree with this so :wave:

Comment (joaojusto) [inline style/js/README.md:65]: It's true that right now we don't have any issue with that but I don't believe we are changing for the sake of changing it, I believe that we are discussing if this form is more readable than the one we are using right now.

Comment (pfac) [inline style/js/README.md:65]: I would pitch in but apparently I'm sane.

**Comment by joaojusto** [comment]
@subvisual/developers bump

**Comment by pfac** [comment]
Fine by me :+1:

**Comment by joaojusto** [comment]
@subvisual/developers I would like to try to close the discussion on this one in the next few days. We still have a long way to go here and I feel this will be forgotten again if we don't merge this soon.

**Comment by gabrielpoca** [comment]
:+1:

File: style/js/README.md [hunk 9]
```diff
@@ -1,3 +1,163 @@
...
+---
+* Leave one space between `if` and `(`.
+
+```javascript
+if (something) {
+  //some work here;
+}
+```
+
+* Prefer this form for multiline if statements.
```
Comment (azevedo) [inline style/js/README.md]: `if` statements

File: style/js/README.md [hunk 10]
```diff
@@ -1,3 +1,163 @@
...
+if (something) {
+  // do stuff
+  // even more stuff
+} else {
+  // some other stuff
+  // real stuff
+}
+```
+
+* Don't include spaces before `(` and function declaration.
```
Comment (azevedo) [inline style/js/README.md]: Do you mean _between_ instead of _before_?

File: style/js/README.md [hunk 11]
```diff
@@ -1,3 +1,163 @@
...
+Guides for programing JavaScript in style
+
+General
+---
+* Use [ESLint](http://eslint.org/)
+* Always use `use strict`
+
+Parens and Braces
+---
+* Leave one space between `if` and `(`.
```
Comment (azevedo) [inline style/js/README.md:13]: We should be consistent about appending a _period_ mark at the end of a bullet, or not.

File: style/js/README.md [hunk 12]
```diff
@@ -1,3 +1,163 @@
...
+```javascript
+var map = { ready: 9, when: 4, "you are": 15 };
+```
+
+* Prefer single line object/array initialization.
+
+```javascript
+var map = { when: 4, ready: 9, "you are": 15 };
+
+var map = [1, 2, 3];
```
Comment (azevedo) [inline style/js/README.md]: I think we should use a different modifier here since we've used it in the line before

File: style/js/README.md [hunk 13]
```diff
@@ -1,3 +1,163 @@
...
+---
+
+* Prefer this form for constructors.
+
+```javascript
+function Foo(bar) {
+  this.bar = bar;
+}
+
+Foo.prototype.getBar = function() {
```
Comment (azevedo) [inline style/js/README.md:83]: What's our guidelines for having named functions instead of anonymous. The difference here would be a better debugging experience

Comment (gabrielpoca) [inline style/js/README.md:83]: There is no need for that with ES6. I think this whole section should be updated. But I do believe we should merge this as soon as possible, and open individual pull requests for more changes, it'll be easier to resolve smaller conversations.

Comment (azevedo) [inline style/js/README.md:83]: I know that, but this guide has no reference to ES6 and we're not using it in all projects

Comment (gabrielpoca) [inline style/js/README.md:83]: But that's the thing, we should merge this and update for es6. Maybe we aren't using es6 in all projects, but I guess we'll be using from now on.

Comment (azevedo) [inline style/js/README.md:83]: Hopefully, but can't be certain.

Comment (gabrielpoca) [inline style/js/README.md:83]: Why not? I believe the guides should reflect what we ought to do from now on.

Comment (azevedo) [inline style/js/README.md:83]: For instance, I'm not sure the Rails environment is not yet friendly with es6. But I'm all for going with es6 everywhere

Comment (gabrielpoca) [inline style/js/README.md:83]: :+1:

**Comment by joaojusto** [comment]
@azevedo-252 updated to address your comments.

File: style/js/README.md (hunk 13 above)
Comment (naps62) [inline style/js/README.md:83]: We're using blue (which is es6) in the blog (which is rails). There might have been some quirks at the time, regarding file extensions, but it worked perfectly fine otherwise

**Comment by azevedo** [comment]
:+1:
